File: Data Structure/Union find/Hard/LET_no305_Number of Islands II.py
class UnionFind:
    def __init__(self):
        self.parent = dict()
        self.count = 0
        
    def add(self, i, j):
        self.parent[(i, j)] = (i, j)
        self.count += 1
    
    def find(self, i, j):
        if self.parent[(i, j)] != (i, j):
            i2, j2 = self.parent[(i, j)]
            self.parent[(i, j)] = self.find(i2, j2)
        return self.parent[(i, j)]
    
    def union(self, i, j, i2, j2):
        p1i, p1j = self.find(i, j)
        p2i, p2j = self.find(i2, j2)
        if (p1i, p1j) != (p2i, p2j):
            self.count -= 1
            self.parent[(p2i, p2j)] = (p1i, p1j)

class Solution:
    def numIslands2(self, m: int, n: int, positions: List[List[int]]) -> List[int]:
        uf = UnionFind()
        res = []
        
        for i, j in positions:
            if (i, j) not in uf.parent:
                uf.add(i, j)
                for di, dj in [(0, 1), (1, 0), (-1, 0), (0, -1)]:
                    i2, j2 = i + di, j + dj
                    if 0 <= i2 < m and 0 <= j2 < n and (i2, j2) in uf.parent:
                        uf.union(i, j, i2, j2)
            
            res.append(uf.count)
        
        return res

File: Data Structure/Union find/Hard/LET_no305_Number of Islands II.cpp
class UnionFind {
private:
    int count, m, n;
    vector<int> parent;
public:
    UnionFind(int m, int n) { 
        count = 0;
        this->m = m;
        this->n = n;
        parent.resize(m * n, -1);
    }
    
    bool hasParent(int i, int j) {
        int idx = i * n + j;
        return parent[idx] != -1;
    }

    void add(int i, int j) {
        int idx = i * n + j;
        parent[idx] = idx;
        count++;
    }

    int getParent(int i, int j) {
        int idx = i * n + j;

        if (parent[idx] != idx) {
            i = parent[idx] / n;
            j = parent[idx] % n;
            parent[idx] = getParent(i, j);
        }

        return parent[idx];
    }

    void unite(int i, int j, int i2, int j2) {
        int p1 = getParent(i, j);
        int p2 = getParent(i2, j2);
        if (p1 != p2) {
            count--;
            parent[p1] = p2;
        }
    }

    int getCount() { return count; }
};

class Solution {
public:
    vector<int> numIslands2(int m, int n, vector<vector<int>>& positions) {
        UnionFind uf(m, n);
        vector<int> res;
        vector<vector<int>> shift = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        int i, j, i2, j2;

        for (auto& pos : positions) {
            i = pos[0];
            j = pos[1];
            if (!uf.hasParent(i, j)) {
                uf.add(i, j);
                for (auto& s : shift) {
                    i2 = i + s[0];
                    j2 = j + s[1];
                    if (0 <= i2 && i2 < m && 0 <= j2 && j2 < n && uf.hasParent(i2, j2))
                        uf.unite(i, j, i2, j2);
                }
            }

            res.push_back(uf.getCount());
        }

        return res;
    }
};